### PR TITLE
Add ONNX Support for Decision Transformer Model

### LINKED ISSUE #2032: ONNX support for decision transformers
### Feature request

I am trying to train off-line RL using decision  transformer, convert to .onnx.

```
from pathlib import Path
from transformers.onnx import FeaturesManager

feature = "sequence-classification"

# load config
model_kind, model_onnx_config = FeaturesManager.check_supported_model_or_raise(model, feature=feature)
onnx_config = model_onnx_config(model.config)

# export
onnx_inputs, onnx_outputs = transformers.onnx.export(
        #preprocessor=tokenizer,
        model=model,
        config=onnx_config,
        opset=13,
        output=Path("trained_models/DT-model.onnx")
)
```

Get the below error:

```
KeyError: "decision-transformer is not supported yet. Only ['albert', 'bart', 'beit', 'bert', 'big-bird', 'bigbird-pegasus', 'blenderbot', 'blenderbot-small', 'bloom', 'camembert', 'clip', 'codegen', 'convbert', 'convnext', 'data2vec-text', 'data2vec-vision', 'deberta', 'deberta-v2', 'deit', 'detr', 'distilbert', 'electra', 'flaubert', 'gpt2', 'gptj', 'gpt-neo', 'groupvit', 'ibert', 'imagegpt', 'layoutlm', 'layoutlmv3', 'levit', 'longt5', 'longformer', 'marian', 'mbart', 'mobilebert', 'mobilenet-v1', 'mobilenet-v2', 'mobilevit', 'mt5', 'm2m-100', 'owlvit', 'perceiver', 'poolformer', 'rembert', 'resnet', 'roberta', 'roformer', 'segformer', 'squeezebert', 'swin', 't5', 'vision-encoder-decoder', 'vit', 'whisper', 'xlm', 'xlm-roberta', 'yolos'] are supported. If you want to support decision-transformer please propose a PR or open up an issue."
```

### Motivation

I would want to use trained models in Godot-RL-Agents. Currently agents are trained using PPO OR imitation learning and bothe support onnx format. Supporting decision transformers could hugely help training models navigating complex scenarios.

### Your contribution

I would be interested to raise a PR. But at this time, I have no idea how to go about this. With little bit of guidance, I can try.

### PR DESCRIPTION
**Pull request by ra9hur**
Fixes #2032

@IlyasMoutawwakil ,

As per your suggestions, I have made necessary changes to support decision transformer model to ONNX. I ran the below command:

```
optimum-cli export onnx -m 'edbeeching/decision-transformer-gym-hopper-medium' --task 'reinforcement-learning' --framework 'pt' onnx/
```

Attaching the terminal output:
[output_log.txt](https://github.com/user-attachments/files/17220623/output_log.txt)

The model was exported successfully with a warning:

```
The ONNX export succeeded with the warning: The maximum absolute difference between the output of the reference model and the ONNX exported model is not within the set tolerance 1e-05:
- state_preds: max diff = nan
- action_preds: max diff = nan
- return_preds: max diff = nan
- last_hidden_state: max diff = nan.
```

since `max diff` is `nan`, printed the results of the reference and ONNX models. Have attached the results in the  below text file. 
Note: Have extracted results from output_log.txt file to a separate text file for clarity.
[output.txt](https://github.com/user-attachments/files/17220505/output.txt)

Excluding `nan`, if we closely observe, the difference seems to be well within set tolerance limits.

Please review and let me know your suggestions.

### CODE REVIEW

File: optimum/exporters/onnx/model_configs.py [hunk 1]
```diff
@@ -256,6 +257,23 @@ class ImageGPTOnnxConfig(GPT2OnnxConfig):
...
+    def inputs(self) -> Dict[str, Dict[int, str]]:
+        dynamic_axis = {0: "batch_size", 1: "sequence_length"}
+
+        return {
+            'actions': dynamic_axis,
+            'timesteps': dynamic_axis,
+            'attention_mask': dynamic_axis,
+            'returns_to_go': dynamic_axis,
+            'states': dynamic_axis,
+        }
```
Comment (IlyasMoutawwakil) [inline optimum/exporters/onnx/model_configs.py:275]: I think this is wrong, can you investigate what are the actual dimensions of these inputs

File: optimum/exporters/onnx/base.py [hunk 2]
```diff
@@ -173,6 +173,14 @@ class OnnxConfig(ExportConfig, ABC):
...
             }
         ),
+        "reinforcement-learning": OrderedDict(
+            {
+                "state_preds": {0: "batch_size", 1: "sequence_length"},
+                "action_preds": {0: "batch_size", 1: "sequence_length"},
+                "return_preds": {0: "batch_size", 1: "sequence_length"},
+                "last_hidden_state": {0: "batch_size", 1: "sequence_length"},
+            }
+        ),
```
Comment (IlyasMoutawwakil) [inline optimum/exporters/onnx/base.py:183]: i think this is wrong as well, you'll need to investigate what are the actual dimensions and which ones are dynamic

File: optimum/exporters/tasks.py [hunk 3]
```diff
@@ -217,6 +217,7 @@ class TasksManager:
             "multiple-choice": "AutoModelForMultipleChoice",
             "object-detection": "AutoModelForObjectDetection",
             "question-answering": "AutoModelForQuestionAnswering",
+            "reinforcement-learning": "AutoModel",
```
Comment (IlyasMoutawwakil) [inline optimum/exporters/tasks.py:220]: this is not the right class, this is the feature extraction task class, and it's probably why you're getting nans as output

Comment (IlyasMoutawwakil) [reply optimum/exporters/tasks.py:220]: this model should rather be registered with custom architectures, since it doesn't have an auto class family.

File: optimum/exporters/onnx/model_configs.py (hunk 1 above)
Comment (ra9hur) [reply optimum/exporters/onnx/model_configs.py:275]: Input dimensions are mentioned in input_generator.py. 

```
    def generate(self, input_name: str, framework: str = "pt", int_dtype: str = "int64", float_dtype: str = "fp32"):

        if input_name == "states":
            shape = [self.batch_size, self.normalized_config.config.state_dim]
        elif input_name == "actions":
            shape = [self.batch_size, self.normalized_config.config.act_dim]
        elif input_name == 'returns_to_go':
            shape = [self.batch_size, 1]
        elif input_name == 'timesteps':
            shape = [self.normalized_config.config.state_dim, self.batch_size]
            max_value = self.normalized_config.config.max_ep_len
            return self.random_int_tensor(shape=shape, max_value = max_value, framework=framework, dtype=int_dtype)
        elif input_name == "attention_mask":
            shape = [self.batch_size, self.normalized_config.config.state_dim]

        return self.random_float_tensor(shape, min_value=-1., max_value=1., framework=framework, dtype=float_dtype)
```

When I debug, input shapes as mentioned above are correctly cascaded to the decision transformer model.

File: optimum/exporters/onnx/base.py (hunk 2 above)
Comment (ra9hur) [reply optimum/exporters/onnx/base.py:183]: As mentioned in the output logs, output shapes are also matching based on model signatures.

	- Validating ONNX Model output "state_preds":
	- Validating ONNX Model output "state_preds":
	- Validating ONNX Model output "state_preds":
		-[✓] (2, 11, 11) matches (2, 11, 11)
		-[✓] (2, 11, 11) matches (2, 11, 11)
		-[✓] (2, 11, 11) matches (2, 11, 11)
		-[x] values not close enough, max diff: nan (atol: 1e-05)
		-[x] values not close enough, max diff: nan (atol: 1e-05)
		-[x] values not close enough, max diff: nan (atol: 1e-05)
	- Validating ONNX Model output "action_preds":
	- Validating ONNX Model output "action_preds":
	- Validating ONNX Model output "action_preds":
		-[✓] (2, 11, 3) matches (2, 11, 3)
		-[✓] (2, 11, 3) matches (2, 11, 3)
		-[✓] (2, 11, 3) matches (2, 11, 3)
		-[x] values not close enough, max diff: nan (atol: 1e-05)
		-[x] values not close enough, max diff: nan (atol: 1e-05)
		-[x] values not close enough, max diff: nan (atol: 1e-05)
	- Validating ONNX Model output "return_preds":
	- Validating ONNX Model output "return_preds":
	- Validating ONNX Model output "return_preds":
		-[✓] (2, 11, 1) matches (2, 11, 1)
		-[✓] (2, 11, 1) matches (2, 11, 1)
		-[✓] (2, 11, 1) matches (2, 11, 1)
		-[x] values not close enough, max diff: nan (atol: 1e-05)
		-[x] values not close enough, max diff: nan (atol: 1e-05)
		-[x] values not close enough, max diff: nan (atol: 1e-05)
	- Validating ONNX Model output "last_hidden_state":
	- Validating ONNX Model output "last_hidden_state":
	- Validating ONNX Model output "last_hidden_state":
		-[✓] (2, 33, 128) matches (2, 33, 128)
		-[✓] (2, 33, 128) matches (2, 33, 128)
		-[✓] (2, 33, 128) matches (2, 33, 128)
		-[x] values not close enough, max diff: nan (atol: 1e-05)
		-[x] values not close enough, max diff: nan (atol: 1e-05)
		-[x] values not close enough, max diff: nan (atol: 1e-05)

File: optimum/exporters/tasks.py (hunk 3 above)
Comment (ra9hur) [reply optimum/exporters/tasks.py:220]: OK, thanks for the hint. This could be the reason.

**Comment by ra9hur** [comment]
Thanks @IlyasMoutawwakil  for the detail review comments, will check and re-submit again by next week.

File: optimum/exporters/onnx/model_configs.py (hunk 1 above)
Comment (IlyasMoutawwakil) [reply optimum/exporters/onnx/model_configs.py:275]: yes but the dynamic axis are wrong, they should be specified correctly in the generator and in the model config.

File: optimum/exporters/onnx/base.py (hunk 2 above)
Comment (IlyasMoutawwakil) [reply optimum/exporters/onnx/base.py:183]: yes they are right because onnx won't force the dynamic shapes on the model😅 but we need them to be correct for the model to handle dynamic shapes correctly. for example with the input `timesteps`, you specified a dynamic shape with `0: "batch_size", 1: "sequence_length"`, but that's wrong since you yourself defined the shape in the generator as:
```
        elif input_name == 'timesteps':
            shape = [self.normalized_config.config.state_dim, self.batch_size]
```
where batch size is the second axis and not the first (0).

File: optimum/exporters/onnx/model_configs.py (hunk 1 above)
Comment (ra9hur) [reply optimum/exporters/onnx/model_configs.py]: Have made changes for actual dimensions, please review.

File: optimum/exporters/onnx/base.py (hunk 2 above)
Comment (ra9hur) [reply optimum/exporters/onnx/base.py]: Thanks for the review comments. It's updated, please review.

**Comment by ra9hur** [comment]
@IlyasMoutawwakil ,
I have made necessary changes as per your suggestions. 

`optimum-cli` command runs successfully.

```
optimum-cli export onnx -m 'edbeeching/decision-transformer-gym-hopper-medium' --task 'reinforcement-learning' --framework 'pt' onnx/
```

Here is the terminal output for your reference.
[success_output_log.txt](https://github.com/user-attachments/files/17341898/success_output_log.txt)

Please let me know, if it is good for approval.

File: optimum/exporters/tasks.py (hunk 3 above)
Comment (IlyasMoutawwakil) [reply optimum/exporters/tasks.py:220]: oh you're right ! thanks for investigating !

Comment (IlyasMoutawwakil) [reply optimum/exporters/tasks.py:220]: going through the hub, I see that the reinforcement task being attribute to multiple auto models, eg https://huggingface.co/jat-project/jat is used with the AutoModelForCausalLM, so let's make it a tuple, with ("AutoModel",)

File: optimum/exporters/onnx/model_configs.py [hunk 4]
```diff
@@ -256,6 +258,27 @@ class ImageGPTOnnxConfig(GPT2OnnxConfig):
...
+        DEFAULT_DUMMY_SHAPES['states'] = self._normalized_config.config.state_dim
+        DEFAULT_DUMMY_SHAPES['returns'] = 1
+        DEFAULT_DUMMY_SHAPES['last_hidden_state'] = self._normalized_config.config.hidden_size
+
+        return {
+            'states': {0: 'batch_size', 1: 'sequence_length', 2: 'states'},
+            'actions': {0: 'batch_size', 1: 'sequence_length', 2: 'actions'},
+            'returns_to_go': {0: 'batch_size', 1: 'sequence_length', 2: 'returns'},
+            'timesteps': {0: 'batch_size', 1: 'sequence_length'},
+            'attention_mask': {0: 'batch_size', 1: 'sequence_length'},
```
Comment (IlyasMoutawwakil) [inline optimum/exporters/onnx/model_configs.py:278]: use original dimensions names names here as well (`state_dim` and `act_dim` and `max_ep_len`)
naming them states, actions and returns is confusing because those are the names of the output tensors themselves.

File: optimum/exporters/onnx/model_configs.py [hunk 5]
```diff
@@ -256,6 +258,27 @@ class ImageGPTOnnxConfig(GPT2OnnxConfig):
...
+    DUMMY_INPUT_GENERATOR_CLASSES = (
+        DummyDecisionTransformerInputGenerator,
+    )
+
+    @property
+    def inputs(self) -> Dict[str, Dict[int, str]]:
+        DEFAULT_DUMMY_SHAPES['actions'] = self._normalized_config.config.act_dim
+        DEFAULT_DUMMY_SHAPES['states'] = self._normalized_config.config.state_dim
+        DEFAULT_DUMMY_SHAPES['returns'] = 1
+        DEFAULT_DUMMY_SHAPES['last_hidden_state'] = self._normalized_config.config.hidden_size
```
Comment (IlyasMoutawwakil) [inline optimum/exporters/onnx/model_configs.py:271]: why is this needed ?

File: optimum/utils/input_generators.py [hunk 6]
```diff
@@ -507,6 +507,41 @@ class DummyDecoderTextInputGenerator(DummyTextInputGenerator):
...
+        elif input_name == 'returns_to_go':
+            shape = [self.batch_size, self.sequence_length, 1]
+        elif input_name == 'timesteps':
+            shape = [self.batch_size, self.sequence_length]
+            max_value = max_ep_len
+            return self.random_int_tensor(shape=shape, max_value = max_value, framework=framework, dtype=int_dtype)
+        elif input_name == "attention_mask":
+            shape = [self.batch_size, self.sequence_length]
+
+        return self.random_float_tensor(shape, min_value=-2., max_value=2., framework=framework, dtype=float_dtype)
```
Comment (IlyasMoutawwakil) [inline optimum/utils/input_generators.py:542]:
```suggestion

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.act_dim = self.normalized_config.config.act_dim
        self.state_dim = self.normalized_config.config.state_dim
        self.max_ep_len = self.normalized_config.config.max_ep_len

    def generate(self, input_name: str, framework: str = "pt", int_dtype: str = "int64", float_dtype: str = "fp32"):
        if input_name == "states":
            shape = [self.batch_size, self.sequence_length, self.state_dim]
        elif input_name == "actions":
            shape = [self.batch_size, self.sequence_length, self.act_dim]
        elif input_name == 'returns_to_go':
            shape = [self.batch_size, self.sequence_length, 1]
        elif input_name == "attention_mask":
            shape = [self.batch_size, self.sequence_length]
        elif input_name == 'timesteps':
            shape = [self.batch_size, self.sequence_length]
            return self.random_int_tensor(shape=shape, max_value=max_ep_len, framework=framework, dtype=int_dtype)

        return self.random_float_tensor(shape, min_value=-2., max_value=2., framework=framework, dtype=float_dtype)
```

File: optimum/exporters/onnx/model_configs.py (hunk 5 above)
Comment (IlyasMoutawwakil) [reply optimum/exporters/onnx/model_configs.py:271]:
```suggestion
```

File: tests/exporters/exporters_utils.py [hunk 7]
```diff
@@ -67,6 +67,7 @@
     "data2vec-audio": "hf-internal-testing/tiny-random-Data2VecAudioModel",
     "deberta": "hf-internal-testing/tiny-random-DebertaModel",
     "deberta-v2": "hf-internal-testing/tiny-random-DebertaV2Model",
+    "decision-transformer": "edbeeching/decision-transformer-gym-hopper-medium",
```
Comment (IlyasMoutawwakil) [inline tests/exporters/exporters_utils.py:70]: is there a smaller variant ?

File: optimum/exporters/onnx/base.py [hunk 8]
```diff
@@ -173,6 +173,14 @@ class OnnxConfig(ExportConfig, ABC):
                 "end_logits": {0: "batch_size", 1: "sequence_length"},
             }
         ),
+        "reinforcement-learning": OrderedDict(
+            {
+                "state_preds": {0: "batch_size", 1: "sequence_length", 2: "states"},
+                "action_preds": {0: "batch_size", 1: "sequence_length", 2: "actions"},
+                "return_preds": {0: "batch_size", 1: "sequence_length", 2: "returns"},
+                "last_hidden_state": {0: "batch_size", 1: "sequence_length", 2: "last_hidden_state"},
```
Comment (IlyasMoutawwakil) [inline optimum/exporters/onnx/base.py:181]:
```suggestion
                "return_preds": {0: "batch_size", 1: "sequence_length"},
                "action_preds": {0: "batch_size", 1: "sequence_length", 2: "act_dim"},
                "state_preds": {0: "batch_size", 1: "sequence_length", 2: "state_dim"},
                "last_hidden_state": {0: "batch_size", 1: "sequence_length", 2: "last_hidden_state"},
```

**Comment by IlyasMoutawwakil** [comment]
cool addition ! I committed a couple of suggestions, mainly about keeping the naming of dimensions closer to the config attributes.

File: optimum/utils/input_generators.py [hunk 9]
```diff
@@ -507,6 +507,42 @@ class DummyDecoderTextInputGenerator(DummyTextInputGenerator):
...
+    """
+
+    SUPPORTED_INPUT_NAMES = (
+        'actions',
+        'timesteps',
+        'attention_mask',
+        'returns_to_go',
+        'states',
+    )
+
```
Comment (IlyasMoutawwakil) [inline optimum/utils/input_generators.py:522]:
```suggestion
```

File: tests/exporters/exporters_utils.py (hunk 7 above)
Comment (ra9hur) [reply tests/exporters/exporters_utils.py:70]: There are 3 variants - expert, medium, medium-replay

- 'edbeeching/decision-transformer-gym-hopper-medium'
- 'edbeeching/decision-transformer-gym-hopper-medium-replay'
- 'edbeeching/decision-transformer-gym-hopper-expert'

All are of the same size - 6.6 MB

Comment (IlyasMoutawwakil) [reply tests/exporters/exporters_utils.py:70]: oh okay that's small enough 👍

File: optimum/exporters/onnx/model_configs.py [hunk 10]
```diff
@@ -18,6 +18,7 @@
 from typing import TYPE_CHECKING, Any, Dict, List, Literal, Optional, Tuple, Union
 
 from packaging import version
+from sipbuild.generator.parser.tokens import states
```
Comment (IlyasMoutawwakil) [inline optimum/exporters/onnx/model_configs.py:21]:
```suggestion
```

File: optimum/utils/normalized_config.py [hunk 11]
```diff
@@ -85,6 +85,14 @@ class NormalizedTextConfig(NormalizedConfig):
     EOS_TOKEN_ID = "eos_token_id"
 
 
+class NormalizedDecisionTransformerConfig(NormalizedConfig):
+    # REFERENCE: https://huggingface.co/docs/transformers/model_doc/decision_transformer
+    STATE_DIM = "state_dim"
+    ACT_DIM = "act_dim"
```
Comment (IlyasMoutawwakil) [inline optimum/utils/normalized_config.py:91]:
```suggestion
    ACT_DIM = "act_dim"
    STATE_DIM = "state_dim"

```

File: optimum/utils/input_generators.py [hunk 12]
```diff
@@ -507,6 +507,41 @@ class DummyDecoderTextInputGenerator(DummyTextInputGenerator):
...
+            shape = [self.batch_size, self.sequence_length, self.state_dim]
+        elif input_name == "actions":
+            shape = [self.batch_size, self.sequence_length, self.act_dim]
+        elif input_name == 'returns_to_go':
+            shape = [self.batch_size, self.sequence_length, 1]
+        elif input_name == "attention_mask":
+            shape = [self.batch_size, self.sequence_length]
+        elif input_name == 'timesteps':
+            shape = [self.batch_size, self.sequence_length]
+            return self.random_int_tensor(shape=shape, max_value=max_ep_len, framework=framework, dtype=int_dtype)
```
Comment (IlyasMoutawwakil) [inline optimum/utils/input_generators.py:540]:
```suggestion
            return self.random_int_tensor(shape=shape, max_value=self.max_ep_len, framework=framework, dtype=int_dtype)
```

File: optimum/exporters/tasks.py (hunk 3 above)
Comment (IlyasMoutawwakil) [inline optimum/exporters/tasks.py:220]:
```suggestion
            "reinforcement-learning": ("AutoModel",), # multiple auto model families can be used for reinforcement-learning
```

**Comment by ra9hur** [comment]
@IlyasMoutawwakil ,

Can you please review the changes.

**Comment by IlyasMoutawwakil** [comment]
triggering tests and checking today !

**Comment by ra9hur** [comment]
Sure, thanks @IlyasMoutawwakil 

The branch might be out of sync with the main repo. Please let me know if I should re-sync, copy the PR changes and then commit again.

**Comment by IlyasMoutawwakil** [comment]
@ra9hur yes please, a rebase or merge would be great ! also a `make style` for code quality.

**Comment by ra9hur** [comment]
@IlyasMoutawwakil , I have merged changes in `main` to this feature branch. 

Pardon my ignorance, couldn't quite get what you meant by `make style`. In one of previous commits, you had made changes for `variable names` consistency. Is that what you meant ? Any references so that I can follow ?

**Comment by IlyasMoutawwakil** [comment]
The export seems to fail dues to some outputs mismatch https://github.com/huggingface/optimum/actions/runs/11931260970/job/33256099991?pr=2038#step:5:7057
does it work for you locally ?

> Pardon my ignorance, couldn't quite get what you meant by make style

I mean doing `pip install -e .[quality]` and then `make style` in terminal, so that `ruff` and `black` apply the code quality changes necessary.

**Comment by IlyasMoutawwakil** [comment]
@ra9hur the export does work for `reinforcement-learning` task but not for `feature-extraction`, because the outputs defined for reinforcement learning are specific to this architecture, and it still expects the common single feature extraction output `"feature-extraction": OrderedDict({"last_hidden_state": {0: "batch_size", 1: "sequence_length"}}),`.
I tried changing things directly on your fork, moved the outputs to the model config directly and cleaned up some minor things.
I also ran tests locally:
```
tests/exporters/onnx/test_exporters_onnx_cli.py::OnnxCLIExportTestCase::test_exporters_cli_pytorch_cpu_136_decision_transformer_feature_extraction_default_edbeeching_decision_transformer_gym_hopper_medium PASSED
tests/exporters/onnx/test_exporters_onnx_cli.py::OnnxCLIExportTestCase::test_exporters_cli_pytorch_cpu_137_decision_transformer_no_task_edbeeching_decision_transformer_gym_hopper_medium PASSED
tests/exporters/onnx/test_exporters_onnx_cli.py::OnnxCLIExportTestCase::test_exporters_cli_pytorch_cpu_138_decision_transformer_reinforcement_learning_default_edbeeching_decision_transformer_gym_hopper_medium PASSED
```
we get 3/3 tests that are passing, for both supported tasks and `no_task` which infers it automatically.
tell me if this works for you.

**Comment by ra9hur** [comment]
@IlyasMoutawwakil , I tried  your changes locally and executes without errors. Good to go from my side.
Thanks for helping me to troubleshoot and  fix those issues !!

**Comment by ra9hur** [comment]
@IlyasMoutawwakil , Tried running pytest on the tests folder locally while referring to this link: https://huggingface.co/docs/transformers/testing

`
python -m pytest -n auto --dist=loadfile -s -v ./tests/
`

But, I am not able to exactly reproduce errors as we see here. 
Can you please share how to run run these tests locally ?

**Comment by IlyasMoutawwakil** [comment]
the failing test is only related to the missing hf token, PRs from a fork will usually fail that test for security reasons.
